src/transformers/cache_utils.py

```diff
@@ -0,0 +1,179 @@
+from typing import List, Optional, Tuple
+
+import torch
+
+
+class Cache:
+    def __init__(self) -> None:
```

Comment:
I'd be more in favor of moving all of the following methods:
`__init__`
`__getitem__`
`__iter__`
`__len__`
`get_seq_length`
`to_legacy_cache`
`from_legacy_cache`

 to `DynamicCache` 

because:
- a) We only need backward comp for the dynamic cache and IMO should restrict future Cache implementations to inherit these methods - that would be confusing for new users
- b) IMO `Cache` should a purely abstract object - i.e. it should only have abstract methods that raise `NotImplementedError(...)`
- c) SinkCache should not inherit methods that it then needs to overwrite and call with `super()` such as `get_seq_length` => this is not very readable and makes life hard for people that want to implement their own cache.
- d) I don't think every cache will have the following a list of tuples as `self.key_cache` - when we add the compiled cache we want to have a purely tensor-based key and value cache

Reply:
Makes sense 👍 

It does add a few extra steps in the transition. `generate` will need a few touches (e.g. cache manipulation in beam search), but we can add them along the way.

In the long run, it is indeed better if we go on this route!